src/OrleansRuntime/Streams/QueueBalancer/LeaseBasedQueueBalancer.cs

```diff
@@ -0,0 +1,345 @@
+            }
+        }
+
+        private async Task ReleaseLeasesToMeetResponsibility()
+        {
+            var queueCountToRelease = this.myQueues.Count - this.maximumRespobsibility;
+            if (queueCountToRelease <= 0)
+                return;
+            var queuesToGiveUp = this.myQueues.GetRange(0, queueCountToRelease);
+            await this.leaseProvider.Release(ResourceCategory.Streaming, queuesToGiveUp.Select(queue => queue.AcquiredLease).ToArray());
+            //remove queuesToGiveUp from myQueue list after the balancer released the leases on them
+            this.myQueues.RemoveRange(0, queueCountToRelease);
+            this.logger.Info($"ReleaseLeasesToMeetResponsibility: released {queueCountToRelease} queues, current queue Count: {this.myQueues.Count}");
+        }
+
+        private Task AcquireLeaseToMeetMaxResponsibilty(object state)
+        {
+            return AcquireLeasesToMeetExpectation(this.maximumRespobsibility);
+        }
+
+        private Task AcquireLeasesToMeetMinResponsibility()
+        {
+            return AcquireLeasesToMeetExpectation(this.minimumResponsibilty);
+        }
+
+        private async Task AcquireLeasesToMeetExpectation(int expectedTotalLeaseCount)
+        {
+            int maxAttempts = 5;
+            int attempts = 0;
+            int leasesToAquire = expectedTotalLeaseCount - this.myQueues.Count;
+            this.logger.Info($"AcquireLeasesToMeetExpection : Try to acquire {leasesToAquire} queues");
+            while (attempts ++ <= maxAttempts && leasesToAquire > 0)
+            {
+                //select new queues to acquire
+                var expectedQueues = this.queueSelector.NextSelection(leasesToAquire, this.myQueues.Select(queue=>queue.QueueId)).ToList();
+                var leaseRequests = expectedQueues.Select(queue => new LeaseRequest() {
+                    ResourceKey = queue.ToString(),
+                    Duration = this.leaseLength
+                });
+                var results = await this.leaseProvider.Acquire(ResourceCategory.Streaming, leaseRequests.ToArray());
+                //add successfully acquired queue to myQueues list
+                for (int i = 0; i < results.Count(); i++)
+                {
+                    if (results[i].StatusCode == ResponseCode.OK)
+                    {
+                        this.myQueues.Add(new AcquiredQueue(expectedQueues[i], results[i].AcquiredLease));
+                    }
+                }
+                //if reached expectedTotalLeaseCount
+                if (this.myQueues.Count >= expectedTotalLeaseCount)
+                {
+                    break;
+                }
+            }
+
+            this.logger.Info($"AcquireLeasesToMeetExpection: finished. Now own {this.myQueues.Count} queues. Used attemps : {attempts}, Current minimumReponsibility : {this.minimumResponsibilty}, current maximumResponsibility : {this.maximumRespobsibility}");
```

Comment:
it means acquire leases up to the expected count... is this a bad name? 

A balancer is expected to have a specific amount of leases. get it?